File: src/org/genyris/system/ExitMethod.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.system;

import org.genyris.core.Bignum;
import org.genyris.core.Constants;
import org.genyris.core.Exp;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.AbstractMethod;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;
import org.genyris.interp.UnboundException;

public class ExitMethod extends AbstractMethod {

    public ExitMethod(Interpreter interp) throws GenyrisException {
        super(interp, "exit");
    }

    public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
            throws GenyrisException {
        if( arguments.length < 1 ) {
            System.exit(0); // default ok  exit
            return NIL;
        } else {
            checkArguments(arguments, 1);
            Class[] types = {Bignum.class};
            checkArgumentTypes(types, arguments);
            System.exit(((Bignum)arguments[0]).bigDecimalValue().intValue());
            return NIL;
        }
    }

    public static void bindFunctionsAndMethods(Interpreter interpreter) throws UnboundException, GenyrisException {
        interpreter.bindMethodInstance(Constants.OS, new ExitMethod(interpreter));
    }
}

File: src/org/genyris/os/ExitMethod.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.os

import org.genyris.core.Bignum
import org.genyris.core.Constants
import org.genyris.core.Exp
import org.genyris.exception.GenyrisException
import org.genyris.interp.*

class ExitMethod(interp: Interpreter?) : AbstractMethod(interp, "exit") {
    @kotlin.Throws(GenyrisException::class)
    override fun bindAndExecute(proc: Closure?, arguments: Array<Exp?>, env: Environment?): Exp? {
        var status = 0
        if (arguments.size > 1) {
            checkArguments(arguments, 1)
            val types = arrayOf<Class<*>?>(Bignum::class.java)
            checkArgumentTypes(types, arguments)
            status = (arguments[0] as Bignum).bigDecimalValue().intValue()
            return NIL
        }
        //Runtime.getRuntime().halt(0);
        System.exit(status)
        return NIL
    }

    companion object {
        @kotlin.Throws(UnboundException::class, GenyrisException::class)
        fun bindFunctionsAndMethods(interpreter: Interpreter) {
            interpreter.bindMethodInstance(Constants.OS, ExitMethod(interpreter))
        }
    }
}
